Adding 'Skip' button to notification (like in iTunes)

paullj1 opened the pull request:
Creates 'Skip' button on notifications.

nriley commented:
Nice idea! As this uses SPI, please guard it so it doesn't crash if the SPI is removed, and link to any (reverse-engineered) documentation that may help to explain what you're doing.

Thanks.

paullj1 commented:
I can definitely add some comments to better explain what's going on, but I'm not sure what you mean by SPI.  All of the stuff used is in the Foundation libraries, documented [here](https://developer.apple.com/library/mac/documentation/Foundation/Reference/NSUserNotificationCenter_Class/).

Next commit will contain comments, but for reference:
[NSUserNotification Center Reversed Engineered Docs](https://github.com/indragiek/NSUserNotificationPrivate)

paullj1 commented:
Okay, I see now what you were talking about.  It should be good to go now.

paullj1 commented:
Although, now I'm thinking that it shouldn't matter... since that delegate won't be called if the NSUserNotificationCenter doesn't exist.  Thoughts?

paullj1 commented:
Okay, so I tested it with OS X notifications on/off, as well as with Growler notifications on/off.  The app delegate only gets called when the OS X notifications are on (because the actions are only created in that case).

I also added additional actions if the 'Alert' style notifications are enabled which allow you to thumb up/down (or remove thumb up) from the notification.

Finally, I moved the album art over to the left side of the notification (just like is done in iTunes).  Let me know if you want me to create a new pull request, or merge all of these changes into a single commit.

nriley commented on Sources/Integration/Growler.m:
```diff
@@ -49,17 +49,34 @@ - (void) growl:(Song*)song withImage:(NSData*)image isNew:(BOOL)n {
-    // Notification buttons
+    // Make skip button visible for banner notifications
+    // - see: https://github.com/indragiek/NSUserNotificationPrivate
     [not setValue:@YES forKey:@"_showsButtons"];
```
This is what I was referring to — it should be guarded too.

paullj1 replied:
So I think the 'setValue:forKey' selector is an NSObject method, so I think this will probably still work even if they deprecate this API.  I added a try/catch block that 'gracefully' handles this if the key doesn't exist.  Is that okay?

nriley commented on Sources/Integration/Growler.m:
```diff
@@ -49,17 +49,34 @@ - (void) growl:(Song*)song withImage:(NSData*)image isNew:(BOOL)n {
+    NSUserNotificationAction *likeAction =
+      [NSUserNotificationAction actionWithIdentifier:@"tup" title:likeActionTitle];
+    NSUserNotificationAction *dislikeAction =
+      [NSUserNotificationAction actionWithIdentifier:@"tud" title:@"Thumb Down"];
```
Please use consistent wording with the rest of Hermes (Like/Dislike).

nriley commented on Sources/Integration/Growler.m:
```diff
@@ -49,17 +49,34 @@ - (void) growl:(Song*)song withImage:(NSData*)image isNew:(BOOL)n {
+      [NSUserNotificationAction actionWithIdentifier:@"tud" title:@"Thumb Down"];
+    
+    [not setAdditionalActions:
+      [NSArray arrayWithObjects: skipAction,likeAction,dislikeAction,nil]];
```
You can use Objective-C literal syntax here.

nriley commented on Sources/Integration/Growler.m:
```diff
@@ -115,13 +132,36 @@ - (BOOL)userNotificationCenter:(NSUserNotificationCenter *)center
+  if (actionID != NULL) {
+    
+    // One of the drop down buttons was pressed
+    if ([actionID caseInsensitiveCompare:@"tup"] == NSOrderedSame) {
```
isEqualToString: should be fine here.

nriley commented:
It's fine if this stuff is all in one pull request. Also if this is too much to ask I'm happy to fix it up myself. :-)

paullj1 commented:
No worries, it looks like you've done enough for this project!  I'm happy to help if I can.  Just pushed a try/catch block.  Is that what you were looking for?

paullj1 commented:
Also, I'm just now seeing your other comments.  I'll make those changes as well.

paullj1 commented:
Okay, that should do it.  Made all of the requested changes.  Let me know if you need anything else changed!

nriley commented:
Looks good.  Catching exception for the key-value coding failures works fine — actually probably a better idea than using -respondsToSelector:.

Thanks again!
